<!DOCTYPE html>












  


<html class="theme-next pisces use-motion" lang="en">
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">




  
  
    
    
  <script src="/lib/pace/pace.min.js?v=1.0.2"></script>
  <link rel="stylesheet" href="/lib/pace/pace-theme-minimal.min.css?v=1.0.2">























<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2">

<link rel="stylesheet" href="/css/main.css?v=6.7.0">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=6.7.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=6.7.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=6.7.0">


  <link rel="mask-icon" href="/images/logo.svg?v=6.7.0" color="#222">







<script id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '6.7.0',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":true,"scrollpercent":false,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":true,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta name="description" content="ATAC-seq数据实战12345文章:The landscape of accessible chromatin in mammalian preimplantation embryos. Nature 2016 Jun 30;534(7609):652-7. PMID: 27309802GEO：https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc">
<meta name="keywords" content="Seq,ATAC">
<meta property="og:type" content="article">
<meta property="og:title" content="ATAC_Seq">
<meta property="og:url" content="https://benbenmiao.github.io/2019/08/15/ATAC_Seq/index.html">
<meta property="og:site_name" content="BioInfoTech">
<meta property="og:description" content="ATAC-seq数据实战12345文章:The landscape of accessible chromatin in mammalian preimplantation embryos. Nature 2016 Jun 30;534(7609):652-7. PMID: 27309802GEO：https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc">
<meta property="og:locale" content="en">
<meta property="og:updated_time" content="2019-07-17T02:08:24.000Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="ATAC_Seq">
<meta name="twitter:description" content="ATAC-seq数据实战12345文章:The landscape of accessible chromatin in mammalian preimplantation embryos. Nature 2016 Jun 30;534(7609):652-7. PMID: 27309802GEO：https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc">



  <link rel="alternate" href="/atom.xml" title="BioInfoTech" type="application/atom+xml">




  <link rel="canonical" href="https://benbenmiao.github.io/2019/08/15/ATAC_Seq/">



<script id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>ATAC_Seq | BioInfoTech</title>
  












  <noscript>
  <style>
  .use-motion .motion-element,
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-title { opacity: initial; }

  .use-motion .logo,
  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope="" itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>
    <a href="https://github.com/BenbenMiao/" class="github-corner" aria-label="View source on GitHub"><svg width="80" height="80" viewbox="0 0 250 250" style="fill:#151513; color:#fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"/><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"/><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"/></svg></a><style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style>
    <header id="header" class="header" itemscope="" itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta custom-logo">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">BioInfoTech</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
      
        <h1 class="site-subtitle" itemprop="description">BenbenMiao</h1>
      
    
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home">

    
    
    
      
    

    
      
    

    <a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i> <br>Home</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-about">

    
    
    
      
    

    
      
    

    <a href="/about/" rel="section"><i class="menu-item-icon fa fa-fw fa-user"></i> <br>About</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">

    
    
    
      
    

    
      
    

    <a href="/tags/" rel="section"><i class="menu-item-icon fa fa-fw fa-tags"></i> <br>Tags<span class="badge">3</span></a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">

    
    
    
      
    

    
      
    

    <a href="/categories/" rel="section"><i class="menu-item-icon fa fa-fw fa-th"></i> <br>Categories<span class="badge">2</span></a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">

    
    
    
      
    

    
      
    

    <a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i> <br>Archives<span class="badge">2</span></a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-schedule">

    
    
    
      
    

    
      
    

    <a href="/schedule/" rel="section"><i class="menu-item-icon fa fa-fw fa-calendar"></i> <br>Schedule</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-sitemap">

    
    
    
      
    

    
      
    

    <a href="/sitemap.xml" rel="section"><i class="menu-item-icon fa fa-fw fa-sitemap"></i> <br>Sitemap</a>

  </li>
        
        
        
          
          <li class="menu-item menu-item-commonweal">

    
    
    
      
    

    
      
    

    <a href="/404/" rel="section"><i class="menu-item-icon fa fa-fw fa-heartbeat"></i> <br>Commonweal 404</a>

  </li>

      
      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br>Search</a>
        </li>
      
    </ul>
  

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off" placeholder="Searching..." spellcheck="false" type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
            

          
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  
    <div class="reading-progress-bar"></div>
  

  <article class="post post-type-normal" itemscope="" itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="https://benbenmiao.github.io/2019/08/15/ATAC_Seq/">

    <span hidden itemprop="author" itemscope="" itemtype="http://schema.org/Person">
      <meta itemprop="name" content="BenbenMiao">
      <meta itemprop="description" content="A biology information technology blog.">
      <meta itemprop="image" content="/images/avatar.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope="" itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="BioInfoTech">
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">ATAC_Seq

              
            
          </h2>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2019-08-15 15:00:00" itemprop="dateCreated datePublished" datetime="2019-08-15T15:00:00+08:00">2019-08-15</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2019-07-17 10:08:24" itemprop="dateModified" datetime="2019-07-17T10:08:24+08:00">2019-07-17</time>
              
            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope="" itemtype="http://schema.org/Thing"><a href="/categories/Seq/" itemprop="url" rel="index"><span itemprop="name">Seq</span></a></span>

                
                
                  , 
                
              
                <span itemprop="about" itemscope="" itemtype="http://schema.org/Thing"><a href="/categories/Seq/Gene/" itemprop="url" rel="index"><span itemprop="name">Gene</span></a></span>

                
                
              
            </span>
          

          
            
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/2019/08/15/ATAC_Seq/#comments" itemprop="discussionUrl">
                
                  <span class="post-comments-count disqus-comment-count" data-disqus-identifier="2019/08/15/ATAC_Seq/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          
             <span id="/2019/08/15/ATAC_Seq/" class="leancloud_visitors" data-flag-title="ATAC_Seq">
               <span class="post-meta-divider">|</span>
               <span class="post-meta-item-icon">
                 <i class="fa fa-eye"></i>
               </span>
               
                 <span class="post-meta-item-text">Views: </span>
               
                 <span class="leancloud-visitors-count"></span>
             </span>
          

          

          
            <div class="post-symbolscount">
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Symbols count in article: </span>
                
                <span title="Symbols count in article">17k</span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time">15 mins.</span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h1 id="ATAC-seq数据实战"><a href="#ATAC-seq数据实战" class="headerlink" title="ATAC-seq数据实战"></a>ATAC-seq数据实战</h1><figure class="highlight groovy"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">文章:The landscape of accessible chromatin <span class="keyword">in</span> mammalian preimplantation embryos. Nature <span class="number">2016</span> Jun <span class="number">30</span>;<span class="number">534</span>(<span class="number">7609</span>):<span class="number">652</span><span class="number">-7.</span> <span class="string">PMID:</span> <span class="number">27309802</span></span><br><span class="line"></span><br><span class="line">GEO：<span class="string">https:</span><span class="comment">//www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE66581</span></span><br><span class="line"></span><br><span class="line">SRA数据库原始测序数据ID： <span class="string">https:</span><span class="comment">//www.ncbi.nlm.nih.gov/sra?term=SRP055881 把下面的内容保存到文件，命名为 srr.list 就可以使用prefetch这个函数来下载。</span></span><br></pre></td></tr></table></figure>
<h2 id="1-下载作者的数据"><a href="#1-下载作者的数据" class="headerlink" title="1.下载作者的数据"></a>1.下载作者的数据</h2><p>1.1 file:config.sra<br><figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="number">2</span>-cell<span class="number">-1</span> SRR2927015</span><br><span class="line"><span class="number">2</span>-cell<span class="number">-2</span> SRR2927016</span><br><span class="line"><span class="number">2</span>-cell<span class="number">-5</span> SRR3545580</span><br><span class="line"><span class="number">2</span>-cell<span class="number">-4</span> SRR2927018</span><br></pre></td></tr></table></figure></p>
<a id="more"></a>
<p>1.2 linux的sra-toolkit程序的prefetch命令下载<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 下载数据 </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> cat srr.list |<span class="keyword">while</span> <span class="built_in">read</span> id;<span class="keyword">do</span> (nohup <span class="variable">$prefetch</span> <span class="variable">$id</span> -X 100G  &amp; );<span class="keyword">done</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 注意组织好自己的项目</span></span></span><br><span class="line">mkdir -p  ~/project/atac/</span><br><span class="line">cd ~/project/atac/</span><br><span class="line">mkdir &#123;sra,raw,clean,align,peaks,motif,qc&#125;</span><br><span class="line">cd sra </span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># vim 或者cat命令创建 srr.list 文件, 里面保存着作为练习使用的4个数据ID </span></span></span><br><span class="line">source activate atac </span><br><span class="line">cat srr.list |while read id;do ( nohup  prefetch $id &amp; );done</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 默认下载目录：~/ncbi/public/sra/ </span></span></span><br><span class="line">ls -lh ~/ncbi/public/sra/</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 下载耗时，自行解决，学员使用现成数据：/public/project/epi/atac/sra </span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 假如提前下载好了数据。</span></span></span><br><span class="line">cd ~/project/atac/ </span><br><span class="line">ln -s /public/project/epi/atac/sra  sra</span><br></pre></td></tr></table></figure></p>
<p>1.3 结果数据<br><figure class="highlight css"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="selector-tag">-rw-r--r--</span> 1 <span class="selector-tag">stu</span> <span class="selector-tag">stu</span> 4<span class="selector-class">.2G</span> <span class="selector-tag">Aug</span> 25 11<span class="selector-pseudo">:10</span> <span class="selector-tag">SRR2927015</span><span class="selector-class">.sra</span></span><br><span class="line"><span class="selector-tag">-rw-r--r--</span> 1 <span class="selector-tag">stu</span> <span class="selector-tag">stu</span> 5<span class="selector-class">.5G</span> <span class="selector-tag">Aug</span> 25 11<span class="selector-pseudo">:13</span> <span class="selector-tag">SRR2927016</span><span class="selector-class">.sra</span></span><br><span class="line"><span class="selector-tag">-rw-r--r--</span> 1 <span class="selector-tag">stu</span> <span class="selector-tag">stu</span> 2<span class="selector-class">.0G</span> <span class="selector-tag">Aug</span> 25 11<span class="selector-pseudo">:12</span> <span class="selector-tag">SRR2927018</span><span class="selector-class">.sra</span></span><br><span class="line"><span class="selector-tag">-rw-r--r--</span> 1 <span class="selector-tag">stu</span> <span class="selector-tag">stu</span> 7<span class="selector-class">.0G</span> <span class="selector-tag">Aug</span> 25 11<span class="selector-pseudo">:13</span> <span class="selector-tag">SRR3545580</span><span class="selector-class">.sra</span></span><br></pre></td></tr></table></figure></p>
<h2 id="2-sra数据转换为fq格式"><a href="#2-sra数据转换为fq格式" class="headerlink" title="2.sra数据转换为fq格式"></a>2.sra数据转换为fq格式</h2><p>2.1 sra数据转换为fq格式<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 下面需要用循环</span></span></span><br><span class="line">cd ~/project/atac/</span><br><span class="line">source activate atac</span><br><span class="line">dump=fastq-dump</span><br><span class="line">analysis_dir=raw</span><br><span class="line">mkdir -p $analysis_dir</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 下面用到的 config.sra 文件，就是上面自行制作的。</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="variable">$fastq</span>-dump sra/SRR2927015.sra  --gzip --split-3  -A 2-cell-1 -O clean/</span></span><br><span class="line">cat config.sra |while read id;</span><br><span class="line">do echo $id</span><br><span class="line">arr=($id)</span><br><span class="line">srr=$&#123;arr[1]&#125;</span><br><span class="line">sample=$&#123;arr[0]&#125;</span><br><span class="line"><span class="meta">#</span><span class="bash">  测序数据的sra转fasq</span></span><br><span class="line">nohup $dump -A  $sample -O $analysis_dir  --gzip --split-3  sra/$srr.sra &amp; </span><br><span class="line">done </span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">## 如果不只是4个文件，需要使用shell脚本批处理。</span></span></span><br><span class="line">cut -f  10,13 SRP055881/SraRunTable.txt|\</span><br><span class="line">sed 's/Embryonic stem cell/ESC/'|sed 's/early 2-cell/e2-cell/' |\</span><br><span class="line">perl -alne '&#123;$h&#123;$F[1]&#125;++;print "$_-$h&#123;$F[1]&#125;"&#125;' |tail -n+2|awk '&#123;print $2"\t"$1&#125;'&gt; config.sra</span><br></pre></td></tr></table></figure></p>
<p>2.2 原始fq数据<br><figure class="highlight css"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 2<span class="selector-class">.6G</span> <span class="selector-tag">Aug</span> 24 23<span class="selector-pseudo">:10</span> 2<span class="selector-tag">-cell-1_1</span><span class="selector-class">.fastq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 2<span class="selector-class">.6G</span> <span class="selector-tag">Aug</span> 24 23<span class="selector-pseudo">:10</span> 2<span class="selector-tag">-cell-1_2</span><span class="selector-class">.fastq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 3<span class="selector-class">.4G</span> <span class="selector-tag">Aug</span> 24 23<span class="selector-pseudo">:31</span> 2<span class="selector-tag">-cell-2_1</span><span class="selector-class">.fastq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 3<span class="selector-class">.7G</span> <span class="selector-tag">Aug</span> 24 23<span class="selector-pseudo">:31</span> 2<span class="selector-tag">-cell-2_2</span><span class="selector-class">.fastq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 1<span class="selector-class">.2G</span> <span class="selector-tag">Aug</span> 24 22<span class="selector-pseudo">:46</span> 2<span class="selector-tag">-cell-4_1</span><span class="selector-class">.fastq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 1<span class="selector-class">.2G</span> <span class="selector-tag">Aug</span> 24 22<span class="selector-pseudo">:46</span> 2<span class="selector-tag">-cell-4_2</span><span class="selector-class">.fastq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 4<span class="selector-class">.4G</span> <span class="selector-tag">Aug</span> 24 23<span class="selector-pseudo">:52</span> 2<span class="selector-tag">-cell-5_1</span><span class="selector-class">.fastq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 4<span class="selector-class">.9G</span> <span class="selector-tag">Aug</span> 24 23<span class="selector-pseudo">:52</span> 2<span class="selector-tag">-cell-5_2</span><span class="selector-class">.fastq</span><span class="selector-class">.gz</span></span><br></pre></td></tr></table></figure></p>
<h2 id="3-测序数据的质量控制"><a href="#3-测序数据的质量控制" class="headerlink" title="3.测序数据的质量控制"></a>3.测序数据的质量控制</h2><p>3.1 trim_galore软件进行过滤，双端测序数据的代码如下:<br>需要自行制作 config.raw 文件， 是3列，第一列占位用，第二列是fq1的地址，第3列是fq2的地址。<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">cd ~/project/atac/</span><br><span class="line">mkdir -p clean </span><br><span class="line">source activate atac  </span><br><span class="line"><span class="meta">#</span><span class="bash"> trim_galore -q 25 --phred33 --length 35 -e 0.1 --stringency 4 --paired -o clean/ raw/2-cell-1_1.fastq.gz raw/2-cell-1_2.fastq.gz</span></span><br><span class="line">cat config.raw  |while read id;</span><br><span class="line">do echo $id</span><br><span class="line">arr=($id)</span><br><span class="line">fq2=$&#123;arr[2]&#125;</span><br><span class="line">fq1=$&#123;arr[1]&#125;</span><br><span class="line">sample=$&#123;arr[0]&#125;</span><br><span class="line">nohup  trim_galore -q 25 --phred33 --length 35 -e 0.1 --stringency 4 --paired -o  clean  $fq1   $fq2  &amp; </span><br><span class="line">done </span><br><span class="line">ps -ef |grep trim</span><br></pre></td></tr></table></figure></p>
<p>3.2 过滤后的fq文件<br><figure class="highlight css"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 2<span class="selector-class">.4G</span> <span class="selector-tag">Aug</span> 25 09<span class="selector-pseudo">:35</span> 2<span class="selector-tag">-cell-1_1_val_1</span><span class="selector-class">.fq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 2<span class="selector-class">.3G</span> <span class="selector-tag">Aug</span> 25 09<span class="selector-pseudo">:35</span> 2<span class="selector-tag">-cell-1_2_val_2</span><span class="selector-class">.fq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 3<span class="selector-class">.1G</span> <span class="selector-tag">Aug</span> 25 10<span class="selector-pseudo">:10</span> 2<span class="selector-tag">-cell-2_1_val_1</span><span class="selector-class">.fq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 3<span class="selector-class">.3G</span> <span class="selector-tag">Aug</span> 25 10<span class="selector-pseudo">:10</span> 2<span class="selector-tag">-cell-2_2_val_2</span><span class="selector-class">.fq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 1<span class="selector-class">.1G</span> <span class="selector-tag">Aug</span> 25 08<span class="selector-pseudo">:52</span> 2<span class="selector-tag">-cell-4_1_val_1</span><span class="selector-class">.fq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 1<span class="selector-class">.1G</span> <span class="selector-tag">Aug</span> 25 08<span class="selector-pseudo">:52</span> 2<span class="selector-tag">-cell-4_2_val_2</span><span class="selector-class">.fq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 3<span class="selector-class">.7G</span> <span class="selector-tag">Aug</span> 25 10<span class="selector-pseudo">:27</span> 2<span class="selector-tag">-cell-5_1_val_1</span><span class="selector-class">.fq</span><span class="selector-class">.gz</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">jmzeng</span> <span class="selector-tag">jmzeng</span> 3<span class="selector-class">.9G</span> <span class="selector-tag">Aug</span> 25 10<span class="selector-pseudo">:27</span> 2<span class="selector-tag">-cell-5_2_val_2</span><span class="selector-class">.fq</span><span class="selector-class">.gz</span></span><br></pre></td></tr></table></figure></p>
<p>3.3 fastqc+multiqc可视化<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">cd ~/project/atac/qc </span><br><span class="line">mkdir -p clean </span><br><span class="line">fastqc -t 5  ../clean/2-cell-*gz -o clean </span><br><span class="line">mkdir -p raw </span><br><span class="line">fastqc -t 5  ../raw/2-cell-*gz -o clean </span><br><span class="line"><span class="meta">#</span><span class="bash"> https://zh.wikipedia.org/wiki/ASCII</span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 还有很多其它工具，比如：</span></span></span><br><span class="line">qualimap='/home/jianmingzeng/biosoft/Qualimap/qualimap_v2.2.1/qualimap'</span><br><span class="line"><span class="meta">$</span><span class="bash">qualimap bamqc --java-mem-size=20G  -bam <span class="variable">$id</span>  -outdir ./</span></span><br></pre></td></tr></table></figure></p>
<h2 id="4-序列比对"><a href="#4-序列比对" class="headerlink" title="4.序列比对"></a>4.序列比对</h2><p>4.1 比对需要的index，看清楚物种，根据对应的软件来构建，这里直接用bowtie2进行比对和统计比对率, 需要提前下载参考基因组然后使用命令构建索引，或者直接就下载索引文件：下载小鼠参考基因组的索引和注释文件, 这里用常用的mm10<br><figure class="highlight awk"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 索引大小为3.2GB， 不建议自己下载基因组构建，可以直接下载索引文件，代码如下：</span></span><br><span class="line">mkdir referece &amp;&amp; cd reference</span><br><span class="line">wget -<span class="number">4</span> -q ftp:<span class="regexp">//</span>ftp.ccb.jhu.edu<span class="regexp">/pub/</span>data<span class="regexp">/bowtie2_indexes/mm</span>10.zip</span><br><span class="line">unzip mm10.zip</span><br></pre></td></tr></table></figure></p>
<p>4.2 解压后的索引<br><figure class="highlight oxygene"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="number">848</span>M Jul  <span class="number">5</span> <span class="number">05</span>:<span class="number">03</span> /<span class="keyword">public</span>/<span class="keyword">reference</span>/<span class="keyword">index</span>/bowtie/mm10.<span class="number">1</span>.bt2</span><br><span class="line"><span class="number">633</span>M Jul  <span class="number">5</span> <span class="number">05</span>:<span class="number">00</span> /<span class="keyword">public</span>/<span class="keyword">reference</span>/<span class="keyword">index</span>/bowtie/mm10.<span class="number">2</span>.bt2</span><br><span class="line"><span class="number">6.0</span>K Jul  <span class="number">5</span> <span class="number">05</span>:<span class="number">05</span> /<span class="keyword">public</span>/<span class="keyword">reference</span>/<span class="keyword">index</span>/bowtie/mm10.<span class="number">3</span>.bt2</span><br><span class="line"><span class="number">633</span>M Jul  <span class="number">5</span> <span class="number">05</span>:<span class="number">05</span> /<span class="keyword">public</span>/<span class="keyword">reference</span>/<span class="keyword">index</span>/bowtie/mm10.<span class="number">4</span>.bt2 </span><br><span class="line"><span class="number">848</span>M Jul  <span class="number">5</span> <span class="number">04</span>:<span class="number">52</span> /<span class="keyword">public</span>/<span class="keyword">reference</span>/<span class="keyword">index</span>/bowtie/mm10.rev.<span class="number">1</span>.bt2</span><br><span class="line"><span class="number">633</span>M Jul  <span class="number">5</span> <span class="number">04</span>:<span class="number">49</span> /<span class="keyword">public</span>/<span class="keyword">reference</span>/<span class="keyword">index</span>/bowtie/mm10.rev.<span class="number">2</span>.bt2</span><br></pre></td></tr></table></figure></p>
<p>4.3 单端测序数据的比对代码<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">zcat ../clean/2-cell-1_1_val_1.fq.gz |head -10000 &gt; test1.fq </span><br><span class="line">zcat ../clean/2-cell-1_2_val_2.fq.gz  |head -10000 &gt; test2.fq</span><br><span class="line">bowtie2 -x /public/reference/index/bowtie/mm10 -1 test1.fq  -2 test2.fq</span><br><span class="line">bowtie2 -x /public/reference/index/bowtie/mm10 -1 test1.fq  -2 test2.fq  -S test.sam</span><br><span class="line">bowtie2 -x /public/reference/index/bowtie/mm10 -1 test1.fq  -2 test2.fq  |samtools sort -@ 5 -O bam -o test.bam -</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 建议抛弃 samtools markdup功能，避免麻烦。</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># https://www.jianshu.com/p/1e6189f641db</span></span></span><br><span class="line">samtools markdup -r  test.bam test.samtools.rmdup.bam </span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 把报错信息在谷歌搜索后，在两个网页上找到了答案。</span></span></span><br><span class="line">https://github.com/samtools/samtools/issues/765</span><br><span class="line">https://www.biostars.org/p/288496/</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># gatk 可以在GitHub下载</span></span></span><br><span class="line">/public/biosoft/GATK/gatk-4.0.3.0/gatk  MarkDuplicates \</span><br><span class="line">-I test.bam -O test.picard.rmdup.bam  --REMOVE_SEQUENCING_DUPLICATES true -M test.log </span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">## picards 被包装在GATK里面：</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">## sambamba 文档： http://lomereiter.github.io/sambamba/docs/sambamba-markdup.html</span></span></span><br><span class="line">conda install -y  sambamba</span><br><span class="line">sambamba --help</span><br><span class="line">sambamba markdup --help</span><br><span class="line">sambamba markdup -r test.bam  test.sambamba.rmdup.bam</span><br><span class="line">samtools flagstat test.sambamba.rmdup.bam</span><br><span class="line">samtools flagstat test.bam</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 接下来只保留两条reads要比对到同一条染色体(Proper paired) ，还有高质量的比对结果(Mapping quality&gt;=30)</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 顺便过滤 线粒体reads</span></span></span><br><span class="line">samtools view -f 2 -q 30  test.sambamba.rmdup.bam |grep -v chrM|wc</span><br><span class="line">samtools view -f 2 -q 30  test.sambamba.rmdup.bam |wc</span><br><span class="line">samtools view -h -f 2 -q 30  test.sambamba.rmdup.bam |grep -v chrM| samtools sort  -O bam  -@ 5 -o - &gt; test.last.bam</span><br><span class="line">bedtools bamtobed -i test.last.bam  &gt; test.bed </span><br><span class="line">ls *.bam  |xargs -i samtools index &#123;&#125;</span><br></pre></td></tr></table></figure></p>
<p>4.4 探索好了整个流程，就可以直接写批处理，代码如下<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line">ls /home/jmzeng/project/atac/clean/*_1.fq.gz &gt; 1</span><br><span class="line">ls /home/jmzeng/project/atac/clean/*_2.fq.gz &gt; 2</span><br><span class="line">ls /home/jmzeng/project/atac/clean/*_2.fq.gz |cut -d"/" -f 7|cut -d"_" -f 1  &gt; 0</span><br><span class="line">paste 0 1 2  &gt; config.clean ## 供mapping使用的配置文件</span><br><span class="line"></span><br><span class="line">cd ~/project/epi/align</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 相对目录需要理解</span></span></span><br><span class="line">bowtie2_index=/public/reference/index/bowtie/mm10</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 一定要搞清楚自己的bowtie2软件安装在哪里，以及自己的索引文件在什么地方！！！</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="built_in">source</span> activate atac </span></span><br><span class="line">cat config.clean |while read id;</span><br><span class="line">do echo $id</span><br><span class="line">arr=($id)</span><br><span class="line">fq2=$&#123;arr[2]&#125;</span><br><span class="line">fq1=$&#123;arr[1]&#125;</span><br><span class="line">sample=$&#123;arr[0]&#125;</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 比对过程15分钟一个样本</span></span></span><br><span class="line">bowtie2  -p 5  --very-sensitive -X 2000 -x  $bowtie2_index -1 $fq1 -2 $fq2 |samtools sort  -O bam  -@ 5 -o - &gt; $&#123;sample&#125;.raw.bam </span><br><span class="line">samtools index $&#123;sample&#125;.raw.bam </span><br><span class="line">bedtools bamtobed -i $&#123;sample&#125;.raw.bam  &gt; $&#123;sample&#125;.raw.bed</span><br><span class="line">samtools flagstat $&#123;sample&#125;.raw.bam  &gt; $&#123;sample&#125;.raw.stat</span><br><span class="line"><span class="meta">#</span><span class="bash"> https://github.com/biod/sambamba/issues/177</span></span><br><span class="line">sambamba markdup --overflow-list-size 600000  --tmpdir='./'  -r $&#123;sample&#125;.raw.bam  $&#123;sample&#125;.rmdup.bam</span><br><span class="line">samtools index   $&#123;sample&#125;.rmdup.bam </span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># ref:https://www.biostars.org/p/170294/ </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Calculate %mtDNA:</span></span></span><br><span class="line">mtReads=$(samtools idxstats  $&#123;sample&#125;.rmdup.bam | grep 'chrM' | cut -f 3)</span><br><span class="line">totalReads=$(samtools idxstats  $&#123;sample&#125;.rmdup.bam | awk '&#123;SUM += $3&#125; END &#123;print SUM&#125;')</span><br><span class="line">echo '==&gt; mtDNA Content:' $(bc &lt;&lt;&lt; "scale=2;100*$mtReads/$totalReads")'%'</span><br><span class="line"></span><br><span class="line">samtools flagstat  $&#123;sample&#125;.rmdup.bam &gt; $&#123;sample&#125;.rmdup.stat</span><br><span class="line">samtools view  -h  -f 2 -q 30    $&#123;sample&#125;.rmdup.bam   |grep -v chrM |samtools sort  -O bam  -@ 5 -o - &gt; $&#123;sample&#125;.last.bam</span><br><span class="line">samtools index   $&#123;sample&#125;.last.bam </span><br><span class="line">samtools flagstat  $&#123;sample&#125;.last.bam &gt; $&#123;sample&#125;.last.stat </span><br><span class="line">bedtools bamtobed -i $&#123;sample&#125;.last.bam  &gt; $&#123;sample&#125;.bed</span><br><span class="line">done</span><br></pre></td></tr></table></figure></p>
<p>bowtie2比对加入了-X 2000 参数，是最大插入片段，宽泛的插入片段范围(10-1000bp)</p>
<p>4.5 第一步得到的bam文件<br><figure class="highlight css"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">stu</span> <span class="selector-tag">stu</span> 3<span class="selector-class">.7G</span> <span class="selector-tag">Aug</span> 25 14<span class="selector-pseudo">:17</span> 2<span class="selector-tag">-cell-1</span><span class="selector-class">.bam</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">stu</span> <span class="selector-tag">stu</span> 4<span class="selector-class">.6G</span> <span class="selector-tag">Aug</span> 25 15<span class="selector-pseudo">:32</span> 2<span class="selector-tag">-cell-2</span><span class="selector-class">.bam</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">stu</span> <span class="selector-tag">stu</span> 1<span class="selector-class">.8G</span> <span class="selector-tag">Aug</span> 25 15<span class="selector-pseudo">:47</span> 2<span class="selector-tag">-cell-4</span><span class="selector-class">.bam</span></span><br><span class="line"><span class="selector-tag">-rw-rw-r--</span> 1 <span class="selector-tag">stu</span> <span class="selector-tag">stu</span> 5<span class="selector-class">.5G</span> <span class="selector-tag">Aug</span> 25 16<span class="selector-pseudo">:49</span> 2<span class="selector-tag">-cell-5</span><span class="selector-class">.bam</span></span><br></pre></td></tr></table></figure></p>
<p>4.6 过滤后的bam文件<br><figure class="highlight css"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">3<span class="selector-class">.7G</span> <span class="selector-tag">Aug</span> 25 21<span class="selector-pseudo">:08</span> 2<span class="selector-tag">-cell-1</span><span class="selector-class">.bam</span></span><br><span class="line">490<span class="selector-tag">M</span> <span class="selector-tag">Aug</span> 25 21<span class="selector-pseudo">:14</span> 2<span class="selector-tag">-cell-1</span><span class="selector-class">.last</span><span class="selector-class">.bam</span></span><br><span class="line">776<span class="selector-tag">M</span> <span class="selector-tag">Aug</span> 25 21<span class="selector-pseudo">:13</span> 2<span class="selector-tag">-cell-1</span><span class="selector-class">.rmdup</span><span class="selector-class">.bam</span></span><br><span class="line">4<span class="selector-class">.6G</span> <span class="selector-tag">Aug</span> 25 23<span class="selector-pseudo">:51</span> 2<span class="selector-tag">-cell-2</span><span class="selector-class">.bam</span></span><br><span class="line">678<span class="selector-tag">M</span> <span class="selector-tag">Aug</span> 25 23<span class="selector-pseudo">:58</span> 2<span class="selector-tag">-cell-2</span><span class="selector-class">.last</span><span class="selector-class">.bam</span></span><br><span class="line">1<span class="selector-class">.1G</span> <span class="selector-tag">Aug</span> 25 23<span class="selector-pseudo">:57</span> 2<span class="selector-tag">-cell-2</span><span class="selector-class">.rmdup</span><span class="selector-class">.bam</span></span><br><span class="line">1<span class="selector-class">.8G</span> <span class="selector-tag">Aug</span> 26 00<span class="selector-pseudo">:41</span> 2<span class="selector-tag">-cell-4</span><span class="selector-class">.bam</span></span><br><span class="line">427<span class="selector-tag">M</span> <span class="selector-tag">Aug</span> 26 00<span class="selector-pseudo">:43</span> 2<span class="selector-tag">-cell-4</span><span class="selector-class">.last</span><span class="selector-class">.bam</span></span><br><span class="line">586<span class="selector-tag">M</span> <span class="selector-tag">Aug</span> 26 00<span class="selector-pseudo">:43</span> 2<span class="selector-tag">-cell-4</span><span class="selector-class">.rmdup</span><span class="selector-class">.bam</span></span><br><span class="line">5<span class="selector-class">.5G</span> <span class="selector-tag">Aug</span> 26 03<span class="selector-pseudo">:26</span> 2<span class="selector-tag">-cell-5</span><span class="selector-class">.bam</span></span><br><span class="line">523<span class="selector-tag">M</span> <span class="selector-tag">Aug</span> 26 03<span class="selector-pseudo">:33</span> 2<span class="selector-tag">-cell-5</span><span class="selector-class">.last</span><span class="selector-class">.bam</span></span><br><span class="line">899<span class="selector-tag">M</span> <span class="selector-tag">Aug</span> 26 03<span class="selector-pseudo">:32</span> 2<span class="selector-tag">-cell-5</span><span class="selector-class">.rmdup</span><span class="selector-class">.bam</span></span><br></pre></td></tr></table></figure></p>
<p>4.7 上述脚本的步骤都可以拆分运行，比如bam文件构建index或者转为bed的:<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ls *.last.bam|xargs -i samtools index &#123;&#125; </span><br><span class="line">ls *.last.bam|while read id;do (bedtools bamtobed -i $id &gt;$&#123;id%%.*&#125;.bed) ;done</span><br><span class="line">ls *.raw.bam|while read id;do (nohup bedtools bamtobed -i $id &gt;$&#123;id%%.*&#125;.raw.bed &amp; ) ;done</span><br></pre></td></tr></table></figure></p>
<p>4.8 最后得到的bed文件<br><figure class="highlight css"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">237<span class="selector-tag">M</span> <span class="selector-tag">Aug</span> 26 08<span class="selector-pseudo">:00</span> 2<span class="selector-tag">-cell-1</span><span class="selector-class">.bed</span></span><br><span class="line">338<span class="selector-tag">M</span> <span class="selector-tag">Aug</span> 26 08<span class="selector-pseudo">:01</span> 2<span class="selector-tag">-cell-2</span><span class="selector-class">.bed</span></span><br><span class="line">203<span class="selector-tag">M</span> <span class="selector-tag">Aug</span> 26 08<span class="selector-pseudo">:01</span> 2<span class="selector-tag">-cell-4</span><span class="selector-class">.bed</span></span><br><span class="line">254<span class="selector-tag">M</span> <span class="selector-tag">Aug</span> 26 08<span class="selector-pseudo">:01</span> 2<span class="selector-tag">-cell-5</span><span class="selector-class">.bed</span></span><br></pre></td></tr></table></figure></p>
<h2 id="5-macs2找peaks"><a href="#5-macs2找peaks" class="headerlink" title="5.macs2找peaks"></a>5.macs2找peaks</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> macs2 callpeak -t 2-cell-1.bed  -g mm --nomodel --<span class="built_in">shift</span> -100 --extsize 200  -n 2-cell-1 --outdir ../peaks/</span></span><br><span class="line">ls *.bed | while read id ;do (macs2 callpeak -t $id  -g mm --nomodel --sHit  -100 --extsize 200  -n $&#123;id%%.*&#125; --outdir ../peaks/) ;done </span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># shell 13问</span></span></span><br></pre></td></tr></table></figure>
<h2 id="6-计算插入片段长度，FRiP值，IDR计算重复情况"><a href="#6-计算插入片段长度，FRiP值，IDR计算重复情况" class="headerlink" title="6.计算插入片段长度，FRiP值，IDR计算重复情况"></a>6.计算插入片段长度，FRiP值，IDR计算重复情况</h2><p>非冗余非线粒体能够比对的fragment、比对率、NRF、PBC1、PBC2、peak数、无核小体区NFR、TSS富集、FRiP 、IDR重复的一致性！</p>
<p>6.1 看 sam文件第9列，在R里面统计绘图,绘图R脚本就可以在批量检验bam文件进行出图<br><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">cmd=commandArgs(trailingOnly=<span class="literal">TRUE</span>); </span><br><span class="line">input=cmd[<span class="number">1</span>]; output=cmd[<span class="number">2</span>]; </span><br><span class="line">a=abs(as.numeric(read.table(input)[,<span class="number">1</span>])); </span><br><span class="line">png(file=output);</span><br><span class="line">hist(a,</span><br><span class="line">main=<span class="string">"Insertion Size distribution"</span>,</span><br><span class="line">ylab=<span class="string">"Read Count"</span>,xlab=<span class="string">"Insert Size"</span>,</span><br><span class="line">xaxt=<span class="string">"n"</span>,</span><br><span class="line">breaks=seq(<span class="number">0</span>,max(a),by=<span class="number">10</span>)</span><br><span class="line">); </span><br><span class="line"></span><br><span class="line">axis(side=<span class="number">1</span>,</span><br><span class="line">at=seq(<span class="number">0</span>,max(a),by=<span class="number">100</span>),</span><br><span class="line">labels=seq(<span class="number">0</span>,max(a),by=<span class="number">100</span>)</span><br><span class="line">);</span><br><span class="line"></span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure></p>
<p>6.2 FRiP值的计算：fraction of reads in called peak regions<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">bedtools intersect -a ../new/2-cell-1.bed -b 2-cell-1_peaks.narrowPeak |wc -l</span><br><span class="line">148928</span><br><span class="line">wc ../new/2-cell-1.bed</span><br><span class="line">5105844</span><br><span class="line">wc ../new/2-cell-1.raw.bed</span><br><span class="line">5105844</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">## 搞清楚 FRiP值具体定义：</span></span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">ls *narrowPeak|while  read id;</span><br><span class="line">do </span><br><span class="line">echo $id</span><br><span class="line">bed=../new/$(basename $id "_peaks.narrowPeak").raw.bed</span><br><span class="line"><span class="meta">#</span><span class="bash">ls  -lh <span class="variable">$bed</span> </span></span><br><span class="line">Reads=$(bedtools intersect -a $bed -b $id |wc -l|awk '&#123;print $1&#125;')</span><br><span class="line">totalReads=$(wc -l $bed|awk '&#123;print $1&#125;')</span><br><span class="line">echo $Reads  $totalReads </span><br><span class="line">echo '==&gt; FRiP value:' $(bc &lt;&lt;&lt; "scale=2;100*$Reads/$totalReads")'%'</span><br><span class="line">done</span><br></pre></td></tr></table></figure></p>
<p>6.3 使用R包看不同peaks文件的overlap情况<br><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span>(<span class="literal">F</span>)&#123;</span><br><span class="line">  options(BioC_mirror=<span class="string">"https://mirrors.ustc.edu.cn/bioc/"</span>) </span><br><span class="line">  options(<span class="string">"repos"</span> = c(CRAN=<span class="string">"https://mirrors.tuna.tsinghua.edu.cn/CRAN/"</span>))</span><br><span class="line">  <span class="keyword">source</span>(<span class="string">"http://bioconductor.org/biocLite.R"</span>) </span><br><span class="line"></span><br><span class="line">  <span class="keyword">library</span>(<span class="string">'BiocInstaller'</span>)</span><br><span class="line">  biocLite(<span class="string">"ChIPpeakAnno"</span>)</span><br><span class="line">  biocLite(<span class="string">"ChIPseeker"</span>)</span><br><span class="line">  </span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">library</span>(ChIPseeker)</span><br><span class="line"><span class="keyword">library</span>(ChIPpeakAnno)</span><br><span class="line">list.files(<span class="string">'project/atac/peaks/'</span>,<span class="string">"*.narrowPeak"</span>)</span><br><span class="line">tmp=lapply(list.files(<span class="string">'project/atac/peaks/'</span>,<span class="string">"*.narrowPeak"</span>),<span class="keyword">function</span>(x)&#123;</span><br><span class="line">  <span class="keyword">return</span>(readPeakFile(file.path(<span class="string">'project/atac/peaks/'</span>, x))) </span><br><span class="line">&#125;)</span><br><span class="line"></span><br><span class="line">ol &lt;- findOverlapsOfPeaks(tmp[[<span class="number">1</span>]],tmp[[<span class="number">2</span>]])</span><br><span class="line">png(<span class="string">'overlapVenn.png'</span>)</span><br><span class="line">makeVennDiagram(ol)</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure></p>
<p>6.4 也可以使用专业软件，IDR 来进行计算出来，同时考虑peaks间的overlap，和富集倍数的一致性 。<br><figure class="highlight vim"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">source</span> activate atac</span><br><span class="line"># 可以用<span class="built_in">search</span>先进行检索</span><br><span class="line">conda <span class="built_in">search</span> idr</span><br><span class="line"><span class="keyword">source</span>  deactivate</span><br><span class="line">## 保证所有的软件都是安装在 wes 这个环境下面</span><br><span class="line">conda  create -n <span class="keyword">py3</span> -<span class="keyword">y</span>   <span class="keyword">python</span>=<span class="number">3</span> idr</span><br><span class="line">conda activate <span class="keyword">py3</span></span><br><span class="line">idr -h </span><br><span class="line">idr --samples  <span class="number">2</span>-cell-<span class="number">1</span>_peaks.narrowPeak <span class="number">2</span>-cell-<span class="number">2</span>_peaks.narrowPeak  --plot</span><br></pre></td></tr></table></figure></p>
<p>6.5 结果如下<br><figure class="highlight dns"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">Initial parameter values: [<span class="number">0.10 1.00</span> <span class="number">0.20 0.50</span>]</span><br><span class="line">Final parameter values: [<span class="number">0.00 1.06</span> <span class="number">0.64 0.87</span>]</span><br><span class="line">Number of reported peaks - <span class="number">5893/5893</span> (<span class="number">100</span>.<span class="number">0</span>%)</span><br><span class="line"></span><br><span class="line">Number of peaks passing IDR cutoff of <span class="number">0</span>.<span class="number">05</span> - <span class="number">674/5893</span> (<span class="number">11</span>.<span class="number">4</span>%)</span><br></pre></td></tr></table></figure></p>
<h2 id="7-deeptools的可视化"><a href="#7-deeptools的可视化" class="headerlink" title="7.deeptools的可视化"></a>7.deeptools的可视化</h2><p>7.1 首先把bam文件转为bw文件<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">cd  ~/project/atac/new</span><br><span class="line">source activate atac</span><br><span class="line"><span class="meta">#</span><span class="bash">ls  *.bam  |xargs -i samtools index &#123;&#125; </span></span><br><span class="line">ls *last.bam |while read id;do</span><br><span class="line">nohup bamCoverage -p 5 --normalizeUsing CPM -b $id -o $&#123;id%%.*&#125;.last.bw &amp; </span><br><span class="line">done </span><br><span class="line"></span><br><span class="line">cd dup </span><br><span class="line">ls  *.bam  |xargs -i samtools index &#123;&#125; </span><br><span class="line">ls *.bam |while read id;do</span><br><span class="line">nohup bamCoverage --normalizeUsing CPM -b $id -o $&#123;id%%.*&#125;.rm.bw &amp; </span><br><span class="line">done</span><br></pre></td></tr></table></figure></p>
<p>7.2 查看TSS附件信号强度：<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># both -R and -S can accept multiple files </span></span></span><br><span class="line">mkdir -p  ~/project/atac/tss</span><br><span class="line">cd   ~/project/atac/tss </span><br><span class="line">source activate atac</span><br><span class="line">computeMatrix reference-point  --referencePoint TSS  -p 15  \</span><br><span class="line">-b 10000 -a 10000    \</span><br><span class="line">-R /public/annotation/CHIPseq/mm10/ucsc.refseq.bed  \</span><br><span class="line">-S ~/project/atac/new/*.bw  \</span><br><span class="line">--skipZeros  -o matrix1_test_TSS.gz  \</span><br><span class="line">--outFileSortedRegions regions1_test_genes.bed</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#     both plotHeatmap and plotProfile will use the output from   computeMatrix</span></span></span><br><span class="line">plotHeatmap -m matrix1_test_TSS.gz  -out test_Heatmap.png</span><br><span class="line">plotHeatmap -m matrix1_test_TSS.gz  -out test_Heatmap.pdf --plotFileFormat pdf  --dpi 720  </span><br><span class="line">plotProfile -m matrix1_test_TSS.gz  -out test_Profile.png</span><br><span class="line">plotProfile -m matrix1_test_TSS.gz  -out test_Profile.pdf --plotFileFormat pdf --perGroup --dpi 720 </span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">## 如果要批处理 ，需要学习好linux命令。</span></span></span><br></pre></td></tr></table></figure></p>
<p>下载 bed文件：<a href="https://genome.ucsc.edu/cgi-bin/hgTables" target="_blank" rel="noopener">https://genome.ucsc.edu/cgi-bin/hgTables</a> 只需要3列坐标格式文件。</p>
<p>7.3 查看基因body的信号强度<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">source activate atac</span><br><span class="line">computeMatrix scale-regions  -p 15  \</span><br><span class="line">-R /public/annotation/CHIPseq/mm10/ucsc.refseq.bed  \</span><br><span class="line">-S ~/project/atac/new/*.bw  \</span><br><span class="line">-b 10000 -a 10000  \</span><br><span class="line">--skipZeros -o matrix1_test_body.gz</span><br><span class="line">plotHeatmap -m matrix1_test_body.gz  -out ExampleHeatmap1.png </span><br><span class="line"></span><br><span class="line">plotHeatmap -m matrix1_test_body.gz  -out test_body_Heatmap.png</span><br><span class="line">plotProfile -m matrix1_test_body.gz  -out test_body_Profile.png</span><br></pre></td></tr></table></figure></p>
<p>ngsplot也是可以的。</p>
<h2 id="8-peaks注释"><a href="#8-peaks注释" class="headerlink" title="8.peaks注释"></a>8.peaks注释</h2><p>8.1 统计peak在promoter，exon，intron和intergenic区域的分布<br><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span>(<span class="literal">F</span>)&#123;</span><br><span class="line">options(BioC_mirror=<span class="string">"https://mirrors.ustc.edu.cn/bioc/"</span>) </span><br><span class="line">options(<span class="string">"repos"</span> = c(CRAN=<span class="string">"https://mirrors.tuna.tsinghua.edu.cn/CRAN/"</span>))</span><br><span class="line"><span class="keyword">source</span>(<span class="string">"http://bioconductor.org/biocLite.R"</span>) </span><br><span class="line"></span><br><span class="line"><span class="keyword">library</span>(<span class="string">'BiocInstaller'</span>)</span><br><span class="line">biocLite(<span class="string">"TxDb.Mmusculus.UCSC.mm10.knownGene"</span>)</span><br><span class="line">biocLite(<span class="string">"org.Mm.eg.db"</span>)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">bedPeaksFile         = <span class="string">'8WG16_summits.bed'</span>; </span><br><span class="line">bedPeaksFile</span><br><span class="line"><span class="comment">## loading packages</span></span><br><span class="line"><span class="keyword">require</span>(ChIPseeker)</span><br><span class="line"><span class="keyword">require</span>(TxDb.Mmusculus.UCSC.mm10.knownGene)</span><br><span class="line">txdb &lt;- TxDb.Mmusculus.UCSC.mm10.knownGene</span><br><span class="line"><span class="keyword">require</span>(clusterProfiler) </span><br><span class="line">peak &lt;- readPeakFile( bedPeaksFile )  </span><br><span class="line">keepChr= !grepl(<span class="string">'_'</span>,seqlevels(peak))</span><br><span class="line">seqlevels(peak, pruning.mode=<span class="string">"coarse"</span>) &lt;- seqlevels(peak)[keepChr]</span><br><span class="line">peakAnno &lt;- annotatePeak(peak, tssRegion=c(-<span class="number">3000</span>, <span class="number">3000</span>), </span><br><span class="line">                         TxDb=txdb, annoDb=<span class="string">"org.Mm.eg.db"</span>) </span><br><span class="line">peakAnno_df &lt;- as.data.frame(peakAnno)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">promoter &lt;- getPromoters(TxDb=txdb, upstream=<span class="number">3000</span>, downstream=<span class="number">3000</span>)</span><br><span class="line">tagMatrix &lt;- getTagMatrix(peak, windows=promoter) </span><br><span class="line"><span class="comment"># 然后查看这些peaks在所有基因的启动子附近的分布情况，热图模式</span></span><br><span class="line">tagHeatmap(tagMatrix, xlim=c(-<span class="number">3000</span>, <span class="number">3000</span>), color=<span class="string">"red"</span>)</span><br><span class="line"><span class="comment"># 然后查看这些peaks在所有基因的启动子附近的分布情况，信号强度曲线图</span></span><br><span class="line">plotAvgProf(tagMatrix, xlim=c(-<span class="number">3000</span>, <span class="number">3000</span>), </span><br><span class="line">            xlab=<span class="string">"Genomic Region (5'-&gt;3')"</span>, ylab = <span class="string">"Read Count Frequency"</span>)</span><br><span class="line">plotAnnoPie(peakAnno)</span><br></pre></td></tr></table></figure></p>
<p>8.2 可以载入IGV看看效果，检测软件找到的peaks是否真的合理，还可以配合rmarkdown来出自动化报告。</p>
<p>8.3 Homer 可以做<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> perl ~/miniconda3/envs/atac/share/homer-4.9.1-5/configureHomer.pl  -install mm10 </span></span><br><span class="line"><span class="meta">#</span><span class="bash"> ln -s /home/jmzeng/miniconda3/envs/chipseq/share/homer-4.9.1-5/data/genomes/ genomes</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> cp  /home/jmzeng/miniconda3/envs/chipseq/share/homer-4.9.1-5/config.txt  /home/stu/miniconda3/envs/atac/share/homer-4.9.1-5/config.txt</span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 保证数据库下载是OK</span></span></span><br><span class="line">ls -lh  ~/miniconda3/envs/atac/share/homer-4.9.1-5/data/genomes  </span><br><span class="line">mkdir -p  ~/project/atac/peaks</span><br><span class="line">source activate atac</span><br><span class="line">cd   ~/project/atac/peaks  </span><br><span class="line">ls *.narrowPeak |while read id;</span><br><span class="line">do </span><br><span class="line">echo $id</span><br><span class="line">awk '&#123;print $4"\t"$1"\t"$2"\t"$3"\t+"&#125;' $id &gt;&#123;id%%.*&#125;.homer_peaks.tmp</span><br><span class="line">annotatePeaks.pl  &#123;id%%.*&#125;.homer_peaks.tmp mm10  1&gt;$&#123;id%%.*&#125;.peakAnn.xls</span><br><span class="line"><span class="meta">  2&gt;</span><span class="bash"><span class="variable">$&#123;id%%.*&#125;</span>.annLog.txt</span></span><br><span class="line">done</span><br></pre></td></tr></table></figure></p>
<p>8.4 Bedtools 也可以做</p>
<h2 id="9-motif寻找及注释"><a href="#9-motif寻找及注释" class="headerlink" title="9.motif寻找及注释"></a>9.motif寻找及注释</h2><p>9.1 Homer 可以做<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">ls -lh  ~/miniconda3/envs/atac/share/homer-4.9.1-5/data/genomes  </span><br><span class="line">mkdir -p  ~/project/atac/motif</span><br><span class="line">cd   ~/project/atac/motif  </span><br><span class="line">source activate atac</span><br><span class="line">ls ../peaks/*.narrowPeak |while read id;</span><br><span class="line">do </span><br><span class="line">file=$(basename $id )</span><br><span class="line">sample=$&#123;file%%.*&#125; </span><br><span class="line">echo $sample </span><br><span class="line">awk '&#123;print $4"\t"$1"\t"$2"\t"$3"\t+"&#125;' $id &gt; $&#123;sample&#125;.homer_peaks.tmp</span><br><span class="line">nohup findMotifsGenome.pl $&#123;sample&#125;.homer_peaks.tmp  mm10 $&#123;sample&#125;_motifDir -len 8,10,12  &amp; </span><br><span class="line">done</span><br></pre></td></tr></table></figure></p>
<p>9.2 meme 也可以做,R包，比如 motifmatchr包 也可以做</p>
<h2 id="10-差异peaks分析"><a href="#10-差异peaks分析" class="headerlink" title="10.差异peaks分析"></a>10.差异peaks分析</h2><p>10.1 diffbind 使用R包DiffBind进行chip-seq差异峰的分析-表观组-生信技能树</p>
<p>10.2 为什么我不用 esATAC,因为这个R包就是包装了前面我们讲解的多个分析步骤。<br><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># https://stackoverflow.com/questions/30738974/rjava-load-error-in-rstudio-r-after-upgrading-to-osx-yosemite</span></span><br><span class="line">options(BioC_mirror=<span class="string">"https://mirrors.ustc.edu.cn/bioc/"</span>) </span><br><span class="line">options(<span class="string">"repos"</span> = c(CRAN=<span class="string">"https://mirrors.tuna.tsinghua.edu.cn/CRAN/"</span>))</span><br><span class="line"><span class="keyword">source</span>(<span class="string">"http://bioconductor.org/biocLite.R"</span>) </span><br><span class="line"></span><br><span class="line"><span class="keyword">library</span>(<span class="string">'BiocInstaller'</span>)</span><br><span class="line">biocLite(<span class="string">"TxDb.Mmusculus.UCSC.mm10.knownGene"</span>)</span><br><span class="line">biocLite(<span class="string">"org.Mm.eg.db"</span>)</span><br><span class="line">biocLite(<span class="string">"esATAC"</span>)</span><br><span class="line">install.packages(<span class="string">'idr'</span>)</span><br><span class="line"><span class="keyword">library</span>(esATAC)</span><br><span class="line">printMap()</span><br></pre></td></tr></table></figure></p>
<p>10.3 多组学整合分析</p>
<p>hello <a href="https://benbenmiao.github.io/2019/08/15/ATAC_Seq/">https://benbenmiao.github.io/2019/08/15/ATAC_Seq/</a> bye</p>

      
    </div>

    
      


    

    
    
    

    

    
      
    
    

    
      <div>
        



  



<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>BenbenMiao</li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    
    <a href="https://benbenmiao.github.io/2019/08/15/ATAC_Seq/" title="ATAC_Seq">https://benbenmiao.github.io/2019/08/15/ATAC_Seq/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fa fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.</li>
</ul>

      </div>
    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Seq/" rel="tag"><i class="fa fa-tag"></i> Seq</a>
          
            <a href="/tags/ATAC/" rel="tag"><i class="fa fa-tag"></i> ATAC</a>
          
        </div>
      

      
      
        <div class="post-widgets">
        
          <div class="wp_rating">
            <div id="wpac-rating"></div>
          </div>
        

        

        
        </div>
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/08/15/ChIP_Seq/" rel="prev" title="AChIP_Seq">
                AChIP_Seq <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>


  </div>


          </div>
          

  
    <div class="comments" id="comments">
      <div id="disqus_thread">
        <noscript>Please enable JavaScript to view the comments powered by Disqus.</noscript>
      </div>
    </div>

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope="" itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image" src="/images/avatar.jpg" alt="BenbenMiao">
            
              <p class="site-author-name" itemprop="name">BenbenMiao</p>
              <p class="site-description motion-element" itemprop="description">A biology information technology blog.</p>
          </div>

          
            <nav class="site-state motion-element">
              
                <div class="site-state-item site-state-posts">
                
                  <a href="/archives/">
                
                    <span class="site-state-item-count">2</span>
                    <span class="site-state-item-name">posts</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-categories">
                  <a href="/categories/index.html">
                    
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">2</span>
                    <span class="site-state-item-name">categories</span>
                  </a>
                </div>
              

              
                
                
                <div class="site-state-item site-state-tags">
                  <a href="/tags/index.html">
                    
                    
                      
                    
                      
                    
                      
                    
                    <span class="site-state-item-count">3</span>
                    <span class="site-state-item-name">tags</span>
                  </a>
                </div>
              
            </nav>
          

          
            <div class="feed-link motion-element">
              <a href="/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://github.com/BenbenMiao" title="GitHub &rarr; https://github.com/BenbenMiao" rel="noopener" target="_blank"><i class="fa fa-fw fa-github"></i>GitHub</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="/m-benben@outlook.com" title="Email &rarr; m-benben@outlook.com"><i class="fa fa-fw fa-envelope"></i>Email</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://benbenmiao.github.io/mdeditor/" title="MarkdownEditor &rarr; https://benbenmiao.github.io/mdeditor/"><i class="fa fa-fw fa-pencil-square-o"></i>MarkdownEditor</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://gitlab.com/benbenmiao" title="GitLab &rarr; https://gitlab.com/benbenmiao" rel="noopener" target="_blank"><i class="fa fa-fw fa-gitlab"></i>GitLab</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://gitee.com/BenbenMiao" title="Gitee &rarr; https://gitee.com/BenbenMiao" rel="noopener" target="_blank"><i class="fa fa-fw fa-code"></i>Gitee</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="http://imbenben.icoc.me/" title="WebSite &rarr; http://imbenben.icoc.me/" rel="noopener" target="_blank"><i class="fa fa-fw fa-ravelry"></i>WebSite</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://bitbucket.org/benbenmiao/" title="Bitbucket &rarr; https://bitbucket.org/benbenmiao/" rel="noopener" target="_blank"><i class="fa fa-fw fa-wpexplorer"></i>Bitbucket</a>
                </span>
              
                <span class="links-of-author-item">
                  
                  
                    
                  
                  
                    
                  
                  <a href="https://benbenmiao.github.io/YueR.com/" title="YueR &rarr; https://benbenmiao.github.io/YueR.com/"><i class="fa fa-fw fa-snowflake-o"></i>YueR</a>
                </span>
              
            </div>
          

          
             <div class="cc-license motion-element" itemprop="license">
              
                
              
              
              
              <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" class="cc-opacity" rel="noopener" target="_blank"><img src="/images/cc-by-nc-sa.svg" alt="Creative Commons"></a>
             </div>
          

          
          

          
            
          
          

        </div>
      </div>

      
      <!--noindex-->
        <div class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
            
            
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#ATAC-seq数据实战"><span class="nav-number">1.</span> <span class="nav-text">ATAC-seq数据实战</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#1-下载作者的数据"><span class="nav-number">1.1.</span> <span class="nav-text">1.下载作者的数据</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-sra数据转换为fq格式"><span class="nav-number">1.2.</span> <span class="nav-text">2.sra数据转换为fq格式</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#3-测序数据的质量控制"><span class="nav-number">1.3.</span> <span class="nav-text">3.测序数据的质量控制</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#4-序列比对"><span class="nav-number">1.4.</span> <span class="nav-text">4.序列比对</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#5-macs2找peaks"><span class="nav-number">1.5.</span> <span class="nav-text">5.macs2找peaks</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#6-计算插入片段长度，FRiP值，IDR计算重复情况"><span class="nav-number">1.6.</span> <span class="nav-text">6.计算插入片段长度，FRiP值，IDR计算重复情况</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#7-deeptools的可视化"><span class="nav-number">1.7.</span> <span class="nav-text">7.deeptools的可视化</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#8-peaks注释"><span class="nav-number">1.8.</span> <span class="nav-text">8.peaks注释</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#9-motif寻找及注释"><span class="nav-number">1.9.</span> <span class="nav-text">9.motif寻找及注释</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#10-差异peaks分析"><span class="nav-number">1.10.</span> <span class="nav-text">10.差异peaks分析</span></a></li></ol></li></ol></div>
            

          </div>
        </div>
      <!--/noindex-->
      

      
        <div class="back-to-top">
          <i class="fa fa-arrow-up"></i>
          
        </div>
      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love" id="animate">
    <i class="fa fa-BenbenMiao"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">BenbenMiao</span>

  
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-area-chart"></i>
    </span>
    
      <span class="post-meta-item-text">Symbols count total: </span>
    
    <span title="Symbols count total">26k</span>
  

  
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-coffee"></i>
    </span>
    
      <span class="post-meta-item-text">Reading time total &asymp;</span>
    
    <span title="Reading time total">23 mins.</span>
  
</div>


  <div class="powered-by">Powered by <a href="https://hexo.io" class="theme-link" rel="noopener" target="_blank">Hexo</a> v3.8.0</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme – <a href="https://theme-next.org" class="theme-link" rel="noopener" target="_blank">NexT.Pisces</a> v6.7.0</div>




        








        
      </div>
    </footer>

    

    

    

    
  </div>

  

<script>
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>














  
    
    
  
  <script color="0,0,0" opacity="1" zindex="-1" count="120" src="/lib/canvas-nest/canvas-nest-nomobile.min.js"></script>











  



  
  <script src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script src="/lib/reading_progress/reading_progress.js"></script>


  


  <script src="/js/src/utils.js?v=6.7.0"></script>

  <script src="/js/src/motion.js?v=6.7.0"></script>



  
  


  <script src="/js/src/affix.js?v=6.7.0"></script>

  <script src="/js/src/schemes/pisces.js?v=6.7.0"></script>



  
  <script src="/js/src/scrollspy.js?v=6.7.0"></script>
<script src="/js/src/post-details.js?v=6.7.0"></script>



  


  <script src="/js/src/bootstrap.js?v=6.7.0"></script>



  

  
    <script id="dsq-count-scr" src="https://.disqus.com/count.js" async></script>
  

  
    <script>
      var disqus_config = function () {
        this.page.url = "https://benbenmiao.github.io/2019/08/15/ATAC_Seq/";
        this.page.identifier = "2019/08/15/ATAC_Seq/";
        this.page.title = 'ATAC_Seq';
        };
      function loadComments () {
        var d = document, s = d.createElement('script');
        s.src = 'https://.disqus.com/embed.js';
        s.setAttribute('data-timestamp', '' + +new Date());
        (d.head || d.body).appendChild(s);
      }
      
        $(function () {
          var offsetTop = $('#comments').offset().top - $(window).height();
          if (offsetTop <= 0) {
            // load directly when there's no a scrollbar
            loadComments();
          } else {
            $(window).on('scroll.disqus_scroll', function () {
              // offsetTop may changes because of manually resizing browser window or lazy loading images.
              var offsetTop = $('#comments').offset().top - $(window).height();
              var scrollTop = $(window).scrollTop();

              // pre-load comments a bit? (margin or anything else)
              if (offsetTop - scrollTop < 60) {
                $(window).off('.disqus_scroll');
                loadComments();
              }
            });
          }
        });
      
    </script>
  





  








  
  
  
  
  <script src="//cdn1.lncld.net/static/js/3.11.1/av-min.js"></script>
  <script src="//unpkg.com/valine/dist/Valine.min.js"></script>
  <script>
    var GUEST = ['nick','mail','link'];
    var guest = 'nick,mail,link';
    guest = guest.split(',').filter(function (item) {
      return GUEST.indexOf(item) > -1;
    });
    new Valine({
      el: '#comments' ,
      verify: false,
      notify: false,
      appId: 'HXXn7zaM1LjdPRj99y0xhCXk-gzGzoHsz',
      appKey: '9yRLrfiqskwrIoARzGJobCeC',
      placeholder: 'Just go go',
      avatar: 'mm',
      meta:guest,
      pageSize: '10' || 10,
      visitor: true
    });
  </script>




  

  <script>
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url).replace(/\/{2,}/g, '/');
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x"></i></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x"></i></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  
  <script>
    (function(){
      var bp = document.createElement('script');
      var curProtocol = window.location.protocol.split(':')[0];
      if (curProtocol === 'https') {
        bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';        
      }
      else {
        bp.src = 'http://push.zhanzhang.baidu.com/push.js';
      }
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(bp, s);
    })();
  </script>


  

  
  

  
  

  


  

  
  <script>
    wpac_init = window.wpac_init || [];
    wpac_init.push({
      widget: 'Rating',
      id: 15808,
      el: 'wpac-rating',
      color: 'fc6423'
    });
    (function() {
      if ('WIDGETPACK_LOADED' in window) return;
      WIDGETPACK_LOADED = true;
      var mc = document.createElement('script');
      mc.type = 'text/javascript';
      mc.async = true;
      mc.src = '//embed.widgetpack.com/widget.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(mc, s.nextSibling);
    })();
  </script>


  

  

  

  
  
  
  <script src="/lib/bookmark/bookmark.min.js?v=1.0"></script>
  <script>
  
    bookmark.scrollToMark('auto', "#more");
  
  </script>


  
  <style>
    .copy-btn {
      display: inline-block;
      padding: 6px 12px;
      font-size: 13px;
      font-weight: 700;
      line-height: 20px;
      color: #333;
      white-space: nowrap;
      vertical-align: middle;
      cursor: pointer;
      background-color: #eee;
      background-image: linear-gradient(#fcfcfc, #eee);
      border: 1px solid #d5d5d5;
      border-radius: 3px;
      user-select: none;
      outline: 0;
    }

    .highlight-wrap .copy-btn {
      transition: opacity .3s ease-in-out;
      opacity: 0;
      padding: 2px 6px;
      position: absolute;
      right: 4px;
      top: 8px;
    }

    .highlight-wrap:hover .copy-btn,
    .highlight-wrap .copy-btn:focus {
      opacity: 1
    }

    .highlight-wrap {
      position: relative;
    }
  </style>
  <script>
    $('.highlight').each(function (i, e) {
      var $wrap = $('<div>').addClass('highlight-wrap');
      $(e).after($wrap);
      $wrap.append($('<button>').addClass('copy-btn').append('Copy').on('click', function (e) {
        var code = $(this).parent().find('.code').find('.line').map(function (i, e) {
          return $(e).text();
        }).toArray().join('\n');
        var ta = document.createElement('textarea');
        var range = document.createRange(); //For Chrome
        var sel = window.getSelection(); //For Chrome
        var yPosition = window.pageYOffset || document.documentElement.scrollTop;
        ta.style.top = yPosition + 'px'; //Prevent page scroll
        ta.style.position = 'absolute';
        ta.style.opacity = '0';
        ta.value = code;
        ta.textContent = code; //For FireFox
        ta.contentEditable = true;
        ta.readOnly = false;
        document.body.appendChild(ta);
        range.selectNode(ta);
        sel.removeAllRanges();
        sel.addRange(range);
        ta.setSelectionRange(0, code.length);
        var result = document.execCommand('copy');
        
          if (result) $(this).text('Copied');
          else $(this).text('Copy failed');
        
        ta.blur(); //For iOS
        $(this).blur();
      })).on('mouseleave', function (e) {
        var $b = $(this).find('.copy-btn');
        setTimeout(function () {
          $b.text('Copy');
        }, 300);
      }).append(e);
    })
  </script>


  
  

  
  

  


</body>
</html>
